src/Microsoft.ML.Data/Data/SchemaDefinition.cs

```diff
@@ -392,37 +402,42 @@ public static SchemaDefinition Create(Type userType, Direction direction = Direc
 
                 InternalSchemaDefinition.GetVectorAndItemType(memberInfo, out bool isVector, out Type dataType);
 
-                PrimitiveDataViewType itemType;
-                var keyAttr = memberInfo.GetCustomAttribute<KeyTypeAttribute>();
-                if (keyAttr != null)
-                {
-                    if (!KeyDataViewType.IsValidDataType(dataType))
-                        throw Contracts.ExceptParam(nameof(userType), "Member {0} marked with KeyType attribute, but does not appear to be a valid kind of data for a key type", memberInfo.Name);
-                    if (keyAttr.KeyCount == null)
-                        itemType = new KeyDataViewType(dataType, dataType.ToMaxInt());
-                    else
-                        itemType = new KeyDataViewType(dataType, keyAttr.KeyCount.Count.GetValueOrDefault());
-                }
-                else
-                    itemType = ColumnTypeExtensions.PrimitiveTypeFromType(dataType);
-
                 // Get the column type.
                 DataViewType columnType;
-                var vectorAttr = memberInfo.GetCustomAttribute<VectorTypeAttribute>();
-                if (vectorAttr != null && !isVector)
-                    throw Contracts.ExceptParam(nameof(userType), $"Member {memberInfo.Name} marked with {nameof(VectorTypeAttribute)}, but does not appear to be a vector type", memberInfo.Name);
-                if (isVector)
+                if (!DataViewTypeManager.Knows(dataType, memberInfo.GetCustomAttributes()))
```

Comment:
We are calling `memberInfo.GetCustomAttributes()` multiple times in this method. We should probably cache it in a local variable.